### PR TITLE
Robot UI software

### PR DESCRIPTION
**Pull request by masonw19**
### Description
This is the software to interact with the robot through the LCD display and rotary encoder interface. 

### Testing Done

Ran the program on a jetson nano and modified different values. Read back values from redis server to see that they were stored. Polled redis server every 10 seconds and values did not get changed back to initial values after they had been modified through the user interface.

### Resolved Issues
#2459 

### Length Justification and Key Files to Review
The files in the `redis-test` folder add a lot of lines and do not need to be reviewed, they are used to start the redis server through python for testing. The PR should be kept as a single PR because each part is required to have the robot user interface software to work.

Key files to review: 
 - `robot_ui.py`
 - `screen.py`
 - `home_screen.py`
 - `menu_screen.py`
 - `chip_and_kick_screen.py`
 - `wheel_screen.py`

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

Dashboard:
Webp net-resizeimage (1)
Menu Screen:
config-screen
Wheels Setting Screen:
wheels
Chip and Kick Setting Screen:
chip-and-kick

### CODE REVIEW

File: src/software/jetson_nano/display/robot_ui.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+sys.path.append("./screens/")
+sys.path.append("./lcd_user_interface/")
+sys.path.append("./rotary_encoder/")
+from lcd_user_interface import LcdDisplay
+from rotary_encoder import RotaryEncoder
+from home_screen import HomeScreen
+from menu_screen import MenuScreen
+from wheels_screen import WheelsScreen
+from chip_and_kick_screen import ChipAndKickScreen
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/robot_ui.py]: you should be able to do this through bazel BUILD files

The sys paths can go away once the `py_library`s are setup and the `deps` field is updated correctly

edit: the imports will then become (after adding to BUILD)

`from software.jetson_nano.display.lcd_user_interface import LcdDispaly`

File: src/software/jetson_nano/display/redis-test/config/redis.conf [hunk 2]
```diff
@@ -0,0 +1,1317 @@
+# Redis configuration file example.
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/redis-test/config/redis.conf]: what is this file for?

File: src/software/jetson_nano/display/screens/screen.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import subprocess
+
+"""
+All new screens will inherit from this class. This class will handles editing variables and maintaining
+the current action the cursor is hovering.
+"""
+
+
+class Screen:
+    def __init__(self, lcd_display, status_codes, actions, action_map, draw_screen):
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/screens/screen.py]: also the comment above could be moved below `class Screen`

nit: constructor missing arguments docs

File: src/software/jetson_nano/display/robot_ui.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+This is the top level class for the robot user interface. It ties together the 
+rotary encoder, lcd display, redis server, and all the screens.
+"""
+
+
+class RobotUi:
+    def __init__(self):
+
+        # Initialize redis server and our redis dictionary
+        self.r = redis.Redis(host="localhost", port=6379, db=0)
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/robot_ui.py]: nit: lets maybe call this `self.redis_client`

we should put localhost and 6379 into  `jetson_nano/display/constants.py`

File: src/software/jetson_nano/display/robot_ui.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    def poll_redis(self):
+        """ Update redis dict every 10 seconds """
+        while not self.shutdown:
+            for key in redis_keys:
+                self.redis_dict[key] = float(self.r.get(key).decode("UTF-8"))
+
+            for _, screen in self.screens.items():
+                if _ != "Menu":
+                    screen.update_values(self.redis_dict)
+            time.sleep(10)
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/robot_ui.py]: nit: lets pass the 10 in as an argument,

and in `start_polling` I think we want to set it to 1 second

File: src/software/jetson_nano/display/screens/screen.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        self.lcd_display = lcd_display
+        self.font_size = 12
+        self.font = ImageFont.truetype(
+            "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf", self.font_size
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/screens/screen.py]: let's copy this .ttf font file into a resources folder and link to that to avoid an absolute dependency

File: src/software/jetson_nano/display/robot_ui.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+from wheels_screen import WheelsScreen
+from chip_and_kick_screen import ChipAndKickScreen
+
+# Pins for Rotary Encoder
+BUTTON_PIN = "DAP4_DOUT"  # BOARD 35, TEGRA_SOC: 'DAP4_FS'
+PIN_1 = "GPIO_PE6"  # BOARD 40, TEGRA_SOC: 'DAP4_DOUT'
+PIN_2 = "DAP4_FS"  # BOARD 33, TEGRA_SOC: 'GPIO_PE6'
+
+# These keys indicate how to handle return values
+status_codes = {"none": 0, "change screen": 1, "edit": 2, "update redis": 3}
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/robot_ui.py]: this might work better as a class, and nit maybe rename to `LCDActions` or `ScreenActions`
~~~
class StatusCodes:
    NONE = 0
    CHANGE_SCREEN = 1
    EDIT_SCREEN = 2
    UPDATE_REDIS = 3
~~~

File: src/software/jetson_nano/display/screens/chip_and_kick_screen.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+PADDING = 6
+BASE_Y = 20
+
+"""
+This screen is used to edit the chip and kick speed settings
+"""
+
+
+class ChipAndKickScreen(Screen):
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/screens/chip_and_kick_screen.py:4]: WheelsScreen and ChipAndKickScreen share a lot of code in common

Do you think its possible to consolidate the functionality into `Screen`

The constructor would take in the settings like:
~~~python
[              
         {
                    "param": "chip speed"
                    "type": double,
                    "delta": 0.5,
                    "redis key": "chip speed",
           },
           {
                    "param": "kick speed",
                    "type": double,
                    "delta": 0.5,
                    "redis key": "kick speed",
             }
]
~~~

All settings will either be bools or doubles, so you could pass in the type name

You could still have this class to do additional functionality based on the state of these settings

We could even make the param name and the redis key the same to make it more concise

Comment (akhilveeraghanta) [reply src/software/jetson_nano/display/screens/chip_and_kick_screen.py:4]: That way adding new screens to the `Menu` page would just be creating instances of this consolidated class

File: src/software/jetson_nano/display/screens/screen.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+        self.lcd_display.show()
+
+    def update_screen(self):
+        """ Draw the updated screen """
+        self.draw_screen()
+        self.lcd_display.show()
+        self.draw_header()
+
+    def on_click(self):
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/screens/screen.py:70]: this function is neat!

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
awesome work!! can't wait to use this

File: src/software/jetson_nano/display/robot_ui.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+            time.sleep(10)
+
+    def stop(self):
+        """ Cleanup the GPIO pins """
+        self.shutdown = True
+        self.rotary_encoder.stop()
+
+
+# For testing
+if __name__ == "__main__":
```
Comment (jonathanlew) [inline src/software/jetson_nano/display/robot_ui.py:145]: should this be pulled into a separate test file?

File: src/software/jetson_nano/display/rotary_encoder/rotary_encoder.py [hunk 11]
```diff
@@ -68,11 +68,14 @@ def __init__(
         self.on_counterclockwise_rotate = on_counterclockwise_rotate
         self.on_click = on_click
 
-        self.transitions_per_rotation = len(STATES)
+        self.transitions_per_rotation = len(STATES) - 1
```
Comment (jonathanlew) [inline src/software/jetson_nano/display/rotary_encoder/rotary_encoder.py:72]: why this change?

File: src/software/jetson_nano/display/screens/chip_and_kick_screen.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+        def draw_screen():
+            """ Wheels Screen Layout """
+            self.lcd_display.prepare()
+
+            cursor = ">"
+            cursor_size = self.font.getsize(cursor)[0]
+            cursor_pos_x = 0
+            if self.curr_action != len(self.actions) - 1:
+                cursor_pos_y = 20 + self.font_size * self.curr_action
```
Comment (jonathanlew) [inline src/software/jetson_nano/display/screens/chip_and_kick_screen.py]: I think I saw 20 earlier in this PR, is it a constant that should be pulled out?

File: src/software/jetson_nano/display/screens/home_screen.py [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+        def draw_screen():
+            """ Home Screen Layout """
+            self.lcd_display.prepare()
+
+            cursor_size = self.font.getsize(">")[0]
+
+            x = cursor_size
+            y = 20
+
+            robot_id_str = "Robot ID"
```
Comment (jonathanlew) [inline src/software/jetson_nano/display/screens/home_screen.py]: could we refactor this code a bit like set up robot id, set up battery voltage, set up capacitor voltage, etc?

File: src/software/jetson_nano/display/screens/wheels_screen.py [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+PADDING = 6
+BASE_Y = 20
+
+"""
+This screen is used to edit wheel speed settings
+"""
+
+
+class WheelsScreen(Screen):
+    def __init__(self, lcd_display, redis_dict, status_codes):
```
Comment (jonathanlew) [inline src/software/jetson_nano/display/screens/wheels_screen.py]: can we have some comments on this function?

File: src/software/jetson_nano/display/screens/screen.py [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+        self.lcd_display.draw.rectangle(
+            (0, 0, self.lcd_display.width, self.font_size + 6),
+            outline="#000000",
+            fill="#ffffff",
+        )
+
+        # Get IP address
+        try:
+            cmd = "hostname -I | cut -d' ' -f1"
+            IP = "  IP: " + subprocess.check_output(cmd, shell=True).decode("utf-8")
```
Comment (jonathanlew) [inline src/software/jetson_nano/display/screens/screen.py]: ip and signal strength seem like nice things to pull into util functions

**Review by jonathanlew — COMMENTED** [review]
Awesome work, can't wait to see it in action! Just got a few comments

File: src/software/jetson_nano/display/robot_ui.py (hunk 10 above)
Comment (akhilveeraghanta) [reply src/software/jetson_nano/display/robot_ui.py:145]: does `py_library` let you run this? 

i.e can you do `bazel  run //software/jetson_nano/display:robot_ui`

File: src/software/jetson_nano/display/robot_ui.py (hunk 5 above)
Comment (masonw19) [reply src/software/jetson_nano/display/robot_ui.py]: done

File: src/software/jetson_nano/display/robot_ui.py (hunk 7 above)
Comment (masonw19) [reply src/software/jetson_nano/display/robot_ui.py]: done

File: src/software/jetson_nano/display/screens/chip_and_kick_screen.py (hunk 8 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/chip_and_kick_screen.py:4]: Each screen has their own `draw_screen()` function which describes the layout of the screen on the LCD display and is pretty custom for each screen so I think it is best to keep the parent and child class relationship.  Additionally, the `Menu` file would become quite long if we had all the screen instances being defined in the same file I think.

However, I have refactored the screen files so that there is no need for all the callback functions and the actions are now in a format that is similar to the one you described. This makes the code more concise and simpler to implement new screens in the future. I think a drawback of this change would be that the callback functions allow more customizability of the actions that could be performed but this wouldn't be necessary if the UI will be used for just simple actions.

File: src/software/jetson_nano/display/screens/screen.py (hunk 3 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/screen.py]: Added constructor docs for all classes.

File: src/software/jetson_nano/display/screens/screen.py (hunk 9 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/screen.py:70]: Thanks!

File: src/software/jetson_nano/display/redis-test/config/redis.conf (hunk 2 above)
Comment (masonw19) [reply src/software/jetson_nano/display/redis-test/config/redis.conf]: For testing the connection with a redis server I searched how to start a redis server in python and it said I needed to add this file. 

Is there another way that I should be testing the redis server?

File: src/software/jetson_nano/display/rotary_encoder/rotary_encoder.py (hunk 11 above)
Comment (masonw19) [reply src/software/jetson_nano/display/rotary_encoder/rotary_encoder.py:72]: I was finding that the fourth state transition was not allows being triggered and this was making the user interface feel very laggy. I'm not exactly sure what the cause is for the fourth transition not allows being triggered, it could be hardware related since my breadboard has gotten very messy since adding the LCD display and lots of wires are crossing over each other which could cause some noise. When the rotary encoder code was first pulled into the master branch I had tested it and I don't recall this issue.

Adding the -1 has made the user interface feel much more smooth.

File: src/software/jetson_nano/display/screens/wheels_screen.py [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+                (x, y), back_right_str, font=self.font, fill="#ffffff"
+            )
+            x += self.font.getsize(back_right_str)[0]
+            self.lcd_display.draw.text(
+                (x, y),
+                str(round(self.actions[BR_INDEX]["value"], 1)),
+                font=self.font,
+                fill="#00ffff",
+            )
+
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/screens/wheels_screen.py]: I think we can turn this into a loop from the `actions` dictionary right? something like this:

~~~
            def draw_actions(actions, cursor, start_x, start_y):                                                                                                                                                                        
                cursor = ">"                                                                                                                                                                                                            
                cursor_size = self.font.getsize(cursor)[0]                                                                                                                                                                              
                x = cursor_size                                                                                                                                                                                                         
                y += self.font_size                                                                                                                                                                                                     
                                                                                                                                                                                                                                        
                for action in actions:                                                                                                                                                                                                  
                                                                                                                                                                                                                                        
                    self.lcd_display.draw.text(                                                                                                                                                                                         
                        (x, y), actions["screen name"],                                                                                                                                                                                 
                        font=self.font, fill="#ffffff"                                                                                                                                                                                  
                    )                                                                                                                                                                                                                   
                                                                                                                                                                                                                                        
                    x += self.font.getsize(action["screen name"])[0]                                                                                                                                                                    
                                                                                                                                                                                                                                        
                    self.lcd_display.draw.text(                                                                                                                                                                                         
                        (x, y), str(round(action["value"], 1)),                                                                                                                                                                         
                        font=self.font, fill="#00ffff",                                                                                                                                                                                 
                    )
~~~

and then it can be shared between all screens that have menus

File: src/software/jetson_nano/display/screens/chip_and_kick_screen.py (hunk 8 above)
Comment (akhilveeraghanta) [reply src/software/jetson_nano/display/screens/chip_and_kick_screen.py:4]: That makes sense! Yeah most of the LCDs actions will set something on redis that will trigger some other binary that interacts with the robot.

File: src/software/jetson_nano/display/screens/chip_and_kick_screen.py [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+            x += (self.font.getsize(kick_str))[0]
+            self.lcd_display.draw.text(
+                (x, y), str(round(self.actions[KICK_INDEX]["value"], 1)), font=self.font, fill="#00ffff"
+            )
+
+            x = cursor_size
+            y = self.lcd_display.height - self.font_size - PADDING
+            self.lcd_display.draw.text(
+                (x, y), "Go to Menu screen", font=self.font, fill="#ffffff",
+            )
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/screens/chip_and_kick_screen.py]: we should pull these colors into `jetson_nano/display/colors.py` and define them as constants

also see below on how we can maybe do this iteratively (by iterating over actions) and sharing that code between the two screens

File: src/software/jetson_nano/display/screens/menu_screen.py [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+                "value": "Wheels",
+                "type": str,
+                "delta": None,
+                "screen action": screen_actions.CHANGE_SCREEN
+            },
+            {
+                "redis key": None,
+                "value": "Chip and Kick",
+                "type": str,
+                "delta": 0.5,
```
Comment (masonw19) [inline src/software/jetson_nano/display/screens/menu_screen.py]: "delta" should be None

File: src/software/jetson_nano/display/robot_ui.py [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+PIN_2 = "DAP4_FS" # BOARD 35, TEGRA_SOC: 'DAP4_FS'
+
+# These keys indicate how to handle return values
+#status_codes = {"none": 0, "change screen": 1, "edit": 2, "update redis": 3}
+class ScreenActions:
+    NONE = 0
+    CHANGE_SCREEN = 1
+    EDIT_SCREEN = 2
+    UPDATE_REDIS = 3
+screen_actions = ScreenActions()
```
Comment (masonw19) [inline src/software/jetson_nano/display/robot_ui.py:28]: There is another instance of `ScreenActions()` on line 69. remove one.

File: src/software/jetson_nano/display/screens/wheels_screen.py (hunk 16 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/wheels_screen.py]: I have implemented this function in the `Screen` class so that it is available for all screens that are made. Child screens are still able to pass a `draw_screen` function if they require a different format such as `HomeScreen`. The `draw_actions` function is currently being used by all the child classes except for `HomeScreen`

File: src/software/jetson_nano/display/rotary_encoder/rotary_encoder.py (hunk 11 above)
Comment (jonathanlew) [reply src/software/jetson_nano/display/rotary_encoder/rotary_encoder.py:72]: interesting, please add a quick comment

File: src/software/jetson_nano/display/screens/home_screen.py (hunk 13 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/home_screen.py]: done

File: src/software/jetson_nano/display/screens/wheels_screen.py (hunk 14 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/wheels_screen.py]: done

File: src/software/jetson_nano/display/screens/chip_and_kick_screen.py (hunk 12 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/chip_and_kick_screen.py]: added a `constants.py` file which holds the constants used.

File: src/software/jetson_nano/display/rotary_encoder/rotary_encoder.py (hunk 11 above)
Comment (masonw19) [reply src/software/jetson_nano/display/rotary_encoder/rotary_encoder.py:72]: added comment

File: src/software/jetson_nano/display/screens/screen.py (hunk 15 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/screen.py]: added a utils.py file with the two functions

File: src/software/jetson_nano/display/screens/screen.py (hunk 6 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/screen.py]: done

File: src/software/jetson_nano/display/screens/chip_and_kick_screen.py (hunk 17 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/chip_and_kick_screen.py]: added a `constants.py` file where I placed the colour definitions. added changes to iterate over actions to display on screen in the `screen.py` file.

File: src/software/jetson_nano/display/redis-test/config/redis.conf (hunk 2 above)
Comment (akhilveeraghanta) [reply src/software/jetson_nano/display/redis-test/config/redis.conf]: actually this is useful for not just tests but also for actually running redis, can you move it to jetson_nano/linux_configs

File: src/software/jetson_nano/display/screens/wheels_screen.py [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+    This screen is used to edit wheel speed settings
+    """
+
+    def __init__(self, lcd_display, redis_dict, screen_actions):
+        """
+        @param lcd_display, an instance of the LcdDisplay class
+        @param redis_dict, a dict of values from redis client to init variables on this screen
+        @param screen_actions, an instance of ScreenActions class
+        """
+        actions = [
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/screens/wheels_screen.py:15]: for all the actions, lets just add [m/s] units in the display string

File: src/software/jetson_nano/display/screens/chip_and_kick_screen.py [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+        """
+        actions = [
+            {
+                "redis key": "chip and kick enable",
+                "value": redis_dict["chip and kick enable"],
+                "type": bool,
+                "delta": None,
+                "screen action": screen_actions.UPDATE_REDIS,
+                "display string": "Set Chip & Kick: ",
+            },
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/screens/chip_and_kick_screen.py]: lets seperate this out to
* chip
* kick

because if people set both kick and chip, depending on which one they select the robot will kick or chip (and the kick and chip speeds can both be non zero)


if we set chip or  kick back to false in redis, will the LCD reflect that?

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
Its very close! Just a couple final comments from me +
* we should remove the `resources` folder and just put everything in there in the display folder
* the imports need to be bazelified, checkout the thunderscope folders for BUILD file examples

**Comment by akhilveeraghanta** [comment]
I think you just need fix formatting + pull from upstream master to fix CI

File: src/software/jetson_nano/display/screens/wheels_screen.py (hunk 20 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/wheels_screen.py:15]: added

File: src/software/jetson_nano/display/screens/chip_and_kick_screen.py (hunk 21 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/chip_and_kick_screen.py]: I have seperated chip and kick. 

Yes, if chip or kick is set to false in redis, LCD will reflect that within a few seconds.

File: src/software/jetson_nano/display/robot_ui.py [hunk 22]
```diff
@@ -0,0 +1,168 @@
+import time, redis
+import digitalio
+import board
+from PIL import Image, ImageDraw, ImageOps
+import adafruit_rgb_display.st7735 as st7735
+
+# TODO: test if bazel build works
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/robot_ui.py]: looks like it does :)

File: src/software/jetson_nano/display/screens/chip_and_kick_screen.py [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+            for i in range(self.len):
+                if self.actions[i]["redis key"] == None:
+                    continue
+
+                if self.actions[i]["type"] == bool:
+                    self.actions[i]["value"] = (
+                        1 if redis_dict[self.actions[i]["redis key"]] else 0
+                    )
+                else:
+                    self.actions[i]["value"] = redis_dict[self.actions[i]["redis key"]]
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/screens/chip_and_kick_screen.py]: this function is identical to the one in `wheel` it looks like? Maybe it can be moved to `screen.py`

File: src/software/jetson_nano/display/screens/screen.py [hunk 24]
```diff
@@ -0,0 +1,172 @@
+from PIL import ImageFont
+
+import software.jetson_nano.display.constants as constants
+from software.jetson_nano.display.utils import get_ip_address, get_signal_strength
+
+SCREEN_TYPE = str
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/display/screens/screen.py]: curious what this is? can u maybe add a comment explaining it?

**Review by akhilveeraghanta — COMMENTED** [review]
final set of comments for me, and it lgtm!! :100:

File: src/software/jetson_nano/display/screens/screen.py (hunk 24 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/screen.py]: When displaying the actions on the screen, most of the items have a 'value' field but when changing screens there is no value field. This was a way to identify if the current action was for changing screens and then we would skip displaying the value. I added a comment and changed the if statement for this check so that it may be a bit more clear and that previous variable is no longer needed. 
image

File: src/software/jetson_nano/display/screens/chip_and_kick_screen.py (hunk 23 above)
Comment (masonw19) [reply src/software/jetson_nano/display/screens/chip_and_kick_screen.py]: Yes, it was identical. I've moved it to `screen.py`

**Review by akhilveeraghanta — APPROVED** [review]
lgtm :tv: :fire:

**Review by jonathanlew — APPROVED** [review]
:window:
